numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1846,6 +1846,16 @@ array_empty_like(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
                                             order, dtype, subok);
     Py_DECREF(prototype);
 
+    /* gh-7854: explicitly fill in Py_None for object dtype,
+       otherwise we get weird data corruption
+    */
```

Comment:
Comment is not too helpful and not formated correctly. In any case, this whole block should probably not be here, but one or two levels deeper in PyArray_NewLikeArray or below. Note that empty *does* explicitly get filled with NULL, so you are filling it twice.